kernel/kbd.c

```diff
@@ -54,23 +54,36 @@ void read_key (void) {
     /* Don't */
     r = scancode[c];
   }
-  /* Put the character in the buffer */
-  buf[buf_head] = r;
-  buf_head = (buf_head + 1) & 255;
-  if (CIRC_SPACE(buf_head, buf_tail, 256) == 0) {
-    /* Buffer's full, so drop least-recent char. */
-    buf_tail = (buf_tail + 1) & 255;
+
+  tcb *t = (tcb *)list_pop_front(&kbd_wait);
+  if (t != 0) {
+    /* Threads are waiting; give the first one in line a character. */
+    if (CIRC_CNT(buf_head, buf_tail, 256) > 0) {
+      panic("Logic error in kbd.c; both threads and characters waiting.");
+    }
+    list_remove(&t->wait_queue);
+    t->saved_registers->rax = r;
+    reschedule_thread(t);
+  } else {
+    /* No threads waiting, buffer the character. */
+    buf[buf_head] = r;
+    buf_head = (buf_head + 1) % 256;
```

Comment:
CIRC_INC()?

Reply:
If you mean CIRC_INC(val, size) = (val + 1) % 256, I'm not convinced that saves much time or is easier to read.